raft/raft.cpp

```diff
@@ -882,12 +889,138 @@ raft::last_quorum()
+                       });
+
+        std::sort(match_indices.begin(), match_indices.end());
+        uint32_t median = match_indices[uint32_t(std::ceil(match_indices.size()/2.0))];
+        result = std::min(result, median);
+    }
+
+    return result;
+}
+
+
+std::list<std::set<bzn::uuid_t>>
+raft::get_active_quorum()
+{
+    // TODO: cache this method when the active quorum is updated
+
+    auto extract_uuid = [](const auto& node_json) -> bzn::uuid_t
+        {
+            return node_json["uuid"].asString();
+        };
+
+    bzn::log_entry log_entry = this->last_quorum();
+    switch(log_entry.entry_type) {
+        case bzn::log_entry_type::single_quorum: {
+            std::set<bzn::uuid_t> result;
+            std::transform(log_entry.msg.begin(), log_entry.msg.end(), std::inserter(result, result.begin()),
+                           extract_uuid);
+
+            return std::list<std::set<bzn::uuid_t>>{result};
+        }
+        case bzn::log_entry_type::joint_quorum: {
+            std::set<bzn::uuid_t> result_old, result_new;
+            std::transform(log_entry.msg["old"].begin(), log_entry.msg["old"].end(),
+                           std::inserter(result_old, result_old.begin()),
+                           extract_uuid);
+            std::transform(log_entry.msg["new"].begin(), log_entry.msg["new"].end(),
+                           std::inserter(result_new, result_new.begin()),
+                           extract_uuid);
+
+            return std::list<std::set<bzn::uuid_t>>{result_old, result_new};
+        }
+        default:
+            throw std::runtime_error("last_quorum gave something that's not a quorum");
+    }
+    return std::list<std::set<bzn::uuid_t>>();
+}
+
+
+bool
+raft::is_majority(const std::set<bzn::uuid_t>& votes)
+{
+    for(const auto& s : this->get_active_quorum())
+    {
+        std::set<bzn::uuid_t> intersect;
+        std::set_intersection(s.begin(), s.end(), votes.begin(), votes.end(),
+                              std::inserter(intersect, intersect.begin()));
+
+        if(intersect.size() * 2 <= s.size())
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+
+bool
+raft::in_quorum(const bzn::uuid_t& uuid)
+{
+    const auto quorums = this->get_active_quorum();
+    for(const auto q : quorums)
+    {
+        if (std::find(q.begin(), q.end(),uuid) != q.end())
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+
+bzn::peers_list_t
+raft::get_all_peers()
+{
+    bzn::peers_list_t result;
+    std::vector<std::reference_wrapper<bzn::message>> all_peers;
+    bzn::log_entry log_entry = this->last_quorum();
+
+    if(log_entry.entry_type == bzn::log_entry_type::single_quorum)
+    {
+        all_peers.emplace_back(log_entry.msg);
+    }
+    else
+    {
+        all_peers.emplace_back(log_entry.msg["old"]);
+        all_peers.emplace_back(log_entry.msg["new"]);
+    }
+
+    for(std::reference_wrapper<bzn::message> jpeers   : all_peers)
```

Comment:
auto should work here?

Reply:
SWEET! Good call I'll make that change and see if it works.